README.md
# ExSchedule

This project provides a way to run tasks in an interval basis.

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `ex_schedule` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ex_schedule, "~> 0.2.0"}
  ]
end
```

## Getting Started

After installation you can use the `ExSchedule` module and define the scheduled tasks using the `schedule` macro.

### Example

```elixir
defmodule YourApp.Schedules.Developer do
  use ExSchedule

  schedule every: {6, :hours} do
    Developer.eat(:pizza)
  end

  schedule every: :hour, first_in: {20, :minutes} do
    Developer.drink(:coffee)
  end
end
```

### Authors

- Dimitrios Zorbas (*[Zorbash](https://github.com/Zorbash)*)
- Luiz Varela (*[LuizVarela](https://github.com/Luizvarela)*)


CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.0] - 2022-01-04
## [0.1.0] - 2018-04-23

### Added

* Initial release


test/ex_schedule_test.exs
defmodule ExScheduleTest do
  use ExUnit.Case, async: false

  import ExUnit.CaptureLog

  defmodule TestSchedule do
    use ExSchedule

    schedule name: :tight_schedule, every: {100, :milliseconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_100ms_called)
      end
    end

    schedule every: :day, first_in: {0, :seconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_day_first_in_0)
      end
    end

    schedule every: :day, first_in: {300, :milliseconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_day_first_in_300ms)
      end
    end

    schedule every: :day, first_in: {0, :seconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, {:namespace, namespace()})
      end
    end

    schedule every: :day, first_in: {0, :seconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_day_max_failures)
      end
    end

    schedule name: :failing, every: {100, :milliseconds}, max_failures: 4 do
      raise "Goodbye cruel world"
    end

    schedule name: :failing_indefinitely, every: {100, :milliseconds} do
      raise "Goodbye cruel world"
    end
  end

  def proxy(pid) do
    receive do
      message ->
        send(pid, message)
        proxy(pid)
    end
  end

  alias TestSchedule, as: Subject

  setup do
    if Process.whereis(:test_schedule), do: Process.unregister(:test_schedule)

    __MODULE__
    |> spawn_link(:proxy, [self()])
    |> Process.register(:schedule_test_proxy)

    on_exit(fn ->
      if Process.whereis(:schedule_test_proxy), do: Process.unregister(:schedule_test_proxy)
      if Process.whereis(:test_schedule), do: Process.unregister(:test_schedule)
    end)
  end

  describe "schedules/0" do
    test "returns the List of defined schedules" do
      expected =
        [
          %{interval: 100, first_in: 0},
          %{interval: 86_400_000, first_in: 0},
          %{interval: 86_400_000, first_in: 300},
          %{interval: 86_400_000, first_in: 0},
          %{interval: 86_400_000, first_in: 0},
          %{interval: 100, first_in: 0},
          %{interval: 100, first_in: 0}
        ]
        |> Enum.reverse()

      assert Subject.schedules() |> Enum.map(&Map.take(&1, [:interval, :first_in])) == expected
    end

    test "sets default :max_failures to :infinity" do
      assert [%{max_failures: :infinity} | _] = Subject.schedules() |> Enum.reverse()
    end
  end

  describe "start_link/1" do
    test "returns a Tuple" do
      name = generate_name(Subject)
      {:ok, scheduler} = Subject.start_link(name: name)

      assert is_pid(scheduler)
    end

    test "does not define handlers for undefined schedules" do
      assert_raise FunctionClauseError, fn -> Subject.handle_task(:undefined) end
    end

    test "defines handlers for all defined schedules" do
      Subject.schedules() |> Enum.map(& &1.id) |> Enum.each(&Subject.handle_task(&1))
    rescue
      e in RuntimeError -> %{message: "Goodbye cruel world"} = e
    end

    test "when name is given it registers the process with it" do
      name = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(name: name)

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == name
    end

    test "when name is not given it registers the process with the module" do
      {:ok, scheduler} = Subject.start_link([])

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == ExScheduleTest.TestSchedule
    end

    test "when name and namespace are given it registers with the name" do
      name = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(name: name, namespace: "namespace")

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == name
    end

    test "when namespace is given it registers with the module and the namespace" do
      namespace = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(namespace: namespace)

      name = :"#{ExScheduleTest.TestSchedule}.#{namespace}"

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == name
    end
  end

  describe "namespace/0" do
    test "returns the supervisor namespace" do
      namespace = generate_name(Subject)

      {:ok, _scheduler} = Subject.start_link(namespace: namespace)

      receive do
        {:namespace, ns} -> assert ns == namespace
      after
        100 -> raise "Did not receive the namespace"
      end
    end

    test "with a string namespace returns the supervisor namespace as an Atom" do
      namespace = generate_name(Subject)

      {:ok, _scheduler} = Subject.start_link(namespace: namespace)

      receive do
        {:namespace, ns} -> assert ns == namespace
      after
        100 -> raise "Did not receive the namespace"
      end
    end
  end

  describe "namespace/1" do
    test "with a PID returns the correct namespace" do
      namespace = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(namespace: namespace)

      assert scheduler |> Subject.namespace() == namespace
    end

    test "with a process name returns the correct namespace" do
      namespace = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(namespace: namespace)
      name = Process.info(scheduler, [:registered_name])[:registered_name]

      assert name |> Subject.namespace() == namespace
    end
  end

  describe "scheduling" do
    test "links a child process for each schedule" do
      namespace = generate_name(Subject)
      {:ok, scheduler} = Subject.start_link(namespace: namespace)

      assert scheduler |> Supervisor.which_children() |> length == Subject.schedules() |> length
    end

    test "schedules in the correct intervals" do
      time =
        :timer.tc(fn ->
          {:ok, _scheduler} = Subject.start_link(namespace: generate_name(Subject))

          receive do
            :every_100ms_called ->
              receive do
                :every_100ms_called ->
                  receive do
                    :every_100ms_called -> :nop
                  after
                    150 -> raise "Expected next run to have completed by now"
                  end
              after
                150 -> raise "Expected next run to have completed by now"
              end
          after
            400 -> raise "Expected all runs to have completed by now, but they didn't"
          end
        end)
        |> elem(0)

      assert_in_delta(time / 1000, 200, 50)
    end

    test "when first_in is set to 0, it performs the first run immediately" do
      namespace = generate_name(Subject)
      {:ok, _scheduler} = Subject.start_link(namespace: namespace)

      receive do
        :every_day_first_in_0 -> :nop
      after
        100 -> raise "Expected first run to have occured by now"
      end
    end
  end

  test "when first_in is set, it schedules the first run in that amount" do
    name = generate_name(Subject)
    {:ok, _scheduler} = Subject.start_link(name: name)

    time =
      :timer.tc(fn ->
        receive do
          :every_day_first_in_300ms -> :nop
        after
          400 -> raise "Expected first run to have occured by now"
        end
      end)
      |> elem(0)

    assert_in_delta(time / 1000, 300, 50)
  end

  describe "introspection" do
    test "tasks keep statistics of runs" do
      {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

      :timer.sleep(200)
      assert {_pid, %{runs: 2}} = scheduler |> find_task(%{name: :tight_schedule})
    end

    test "tasks keep statistics of failures" do
      {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

      :timer.sleep(300)
      assert {_pid, %{failures: 3}} = scheduler |> find_task(%{name: :failing})
    end

    test "tasks keep statistics of time to complete" do
      {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

      :timer.sleep(200)
      {_pid, %{last_completed_in: time}} = scheduler |> find_task(%{name: :tight_schedule})

      assert time > 0.0
    end
  end

  describe "failure handling" do
    test "when a process exceeeds :max_failures it is restarted" do
      capture_log(fn ->
        {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

        {pid, _} = scheduler |> find_task(%{name: :failing})
        assert Process.alive?(pid)

        :timer.sleep(500)

        # The old process is no longer alive
        refute Process.alive?(pid)
        {pid, _} = scheduler |> find_task(%{name: :failing})

        # A new task has been spawned
        assert Process.alive?(pid)
      end)
    end

    test "when a failing process is restarted, an error is logged" do
      log_output =
        capture_log(fn ->
          {:ok, _scheduler} = Subject.start_link(name: generate_name(Subject))
          :timer.sleep(500)
        end)

      assert log_output =~ ~r/scheduled task failed with:/
    end

    test "with unspecified :max_failures it is not restarted" do
      capture_log(fn ->
        {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

        {pid, _} = scheduler |> find_task(%{name: :failing_indefinitely})
        assert Process.alive?(pid)

        :timer.sleep(500)

        {^pid, _} = scheduler |> find_task(%{name: :failing_indefinitely})
        assert Process.alive?(pid)
      end)
    end
  end

  defp find_task(scheduler, %{name: name}) do
    scheduler
    |> Supervisor.which_children()
    |> Stream.map(fn {_, pid, _, _} -> {pid, :sys.get_state(pid)} end)
    |> Enum.find(fn {_, %{schedule: %{name: schedule_name}}} -> schedule_name == name end)
  end

  def generate_name(module) do
    :"#{module}-#{make_ref() |> inspect()}"
  end
end


test/ex_scheduler_test.exs
defmodule ExSchedulerTest do
  use ExUnit.Case, async: false

  import ExUnit.CaptureLog

  defmodule TestSchedule do
    use ExScheduler

    schedule name: :tight_schedule, every: {100, :milliseconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_100ms_called)
      end
    end

    schedule every: :day, first_in: {0, :seconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_day_first_in_0)
      end
    end

    schedule every: :day, first_in: {300, :milliseconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_day_first_in_300ms)
      end
    end

    schedule every: :day, first_in: {0, :seconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, {:namespace, namespace()})
      end
    end

    schedule every: :day, first_in: {0, :seconds} do
      with proxy <- Process.whereis(:schedule_test_proxy) do
        if proxy, do: send(proxy, :every_day_max_failures)
      end
    end

    schedule name: :failing, every: {100, :milliseconds}, max_failures: 4 do
      raise "Goodbye cruel world"
    end

    schedule name: :failing_indefinitely, every: {100, :milliseconds} do
      raise "Goodbye cruel world"
    end
  end

  def proxy(pid) do
    receive do
      message ->
        send(pid, message)
        proxy(pid)
    end
  end

  alias TestSchedule, as: Subject

  setup do
    if Process.whereis(:test_schedule), do: Process.unregister(:test_schedule)

    __MODULE__
    |> spawn_link(:proxy, [self()])
    |> Process.register(:schedule_test_proxy)

    on_exit(fn ->
      if Process.whereis(:schedule_test_proxy), do: Process.unregister(:schedule_test_proxy)
      if Process.whereis(:test_schedule), do: Process.unregister(:test_schedule)
    end)
  end

  describe "schedules/0" do
    test "returns the List of defined schedules" do
      expected =
        [
          %{interval: 100, first_in: 0},
          %{interval: 86_400_000, first_in: 0},
          %{interval: 86_400_000, first_in: 300},
          %{interval: 86_400_000, first_in: 0},
          %{interval: 86_400_000, first_in: 0},
          %{interval: 100, first_in: 0},
          %{interval: 100, first_in: 0}
        ]
        |> Enum.reverse()

      assert Subject.schedules() |> Enum.map(&Map.take(&1, [:interval, :first_in])) == expected
    end

    test "sets default :max_failures to :infinity" do
      assert [%{max_failures: :infinity} | _] = Subject.schedules() |> Enum.reverse()
    end
  end

  describe "start_link/1" do
    test "returns a Tuple" do
      name = generate_name(Subject)
      {:ok, scheduler} = Subject.start_link(name: name)

      assert is_pid(scheduler)
    end

    test "does not define handlers for undefined schedules" do
      assert_raise FunctionClauseError, fn -> Subject.handle_task(:undefined) end
    end

    test "defines handlers for all defined schedules" do
      Subject.schedules() |> Enum.map(& &1.id) |> Enum.each(&Subject.handle_task(&1))
    rescue
      e in RuntimeError -> %{message: "Goodbye cruel world"} = e
    end

    test "when name is given it registers the process with it" do
      name = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(name: name)

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == name
    end

    test "when name is not given it registers the process with the module" do
      {:ok, scheduler} = Subject.start_link([])

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == ExSchedulerTest.TestSchedule
    end

    test "when name and namespace are given it registers with the name" do
      name = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(name: name, namespace: "namespace")

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == name
    end

    test "when namespace is given it registers with the module and the namespace" do
      namespace = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(namespace: namespace)

      name = :"#{ExSchedulerTest.TestSchedule}.#{namespace}"

      scheduler_name =
        scheduler
        |> Process.info(:registered_name)
        |> elem(1)

      assert scheduler_name == name
    end
  end

  describe "namespace/0" do
    test "returns the supervisor namespace" do
      namespace = generate_name(Subject)

      {:ok, _scheduler} = Subject.start_link(namespace: namespace)

      receive do
        {:namespace, ns} -> assert ns == namespace
      after
        100 -> raise "Did not receive the namespace"
      end
    end

    test "with a string namespace returns the supervisor namespace as an Atom" do
      namespace = generate_name(Subject)

      {:ok, _scheduler} = Subject.start_link(namespace: namespace)

      receive do
        {:namespace, ns} -> assert ns == namespace
      after
        100 -> raise "Did not receive the namespace"
      end
    end
  end

  describe "namespace/1" do
    test "with a PID returns the correct namespace" do
      namespace = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(namespace: namespace)

      assert scheduler |> Subject.namespace() == namespace
    end

    test "with a process name returns the correct namespace" do
      namespace = generate_name(Subject)

      {:ok, scheduler} = Subject.start_link(namespace: namespace)
      name = Process.info(scheduler, [:registered_name])[:registered_name]

      assert name |> Subject.namespace() == namespace
    end
  end

  describe "scheduling" do
    test "links a child process for each schedule" do
      namespace = generate_name(Subject)
      {:ok, scheduler} = Subject.start_link(namespace: namespace)

      assert scheduler |> Supervisor.which_children() |> length == Subject.schedules() |> length
    end

    test "schedules in the correct intervals" do
      time =
        :timer.tc(fn ->
          {:ok, _scheduler} = Subject.start_link(namespace: generate_name(Subject))

          receive do
            :every_100ms_called ->
              receive do
                :every_100ms_called ->
                  receive do
                    :every_100ms_called -> :nop
                  after
                    150 -> raise "Expected next run to have completed by now"
                  end
              after
                150 -> raise "Expected next run to have completed by now"
              end
          after
            400 -> raise "Expected all runs to have completed by now, but they didn't"
          end
        end)
        |> elem(0)

      assert_in_delta(time / 1000, 200, 50)
    end

    test "when first_in is set to 0, it performs the first run immediately" do
      namespace = generate_name(Subject)
      {:ok, _scheduler} = Subject.start_link(namespace: namespace)

      receive do
        :every_day_first_in_0 -> :nop
      after
        100 -> raise "Expected first run to have occured by now"
      end
    end
  end

  test "when first_in is set, it schedules the first run in that amount" do
    name = generate_name(Subject)
    {:ok, _scheduler} = Subject.start_link(name: name)

    time =
      :timer.tc(fn ->
        receive do
          :every_day_first_in_300ms -> :nop
        after
          400 -> raise "Expected first run to have occured by now"
        end
      end)
      |> elem(0)

    assert_in_delta(time / 1000, 300, 50)
  end

  describe "introspection" do
    test "tasks keep statistics of runs" do
      {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

      :timer.sleep(200)
      assert {_pid, %{runs: 2}} = scheduler |> find_task(%{name: :tight_schedule})
    end

    test "tasks keep statistics of failures" do
      {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

      :timer.sleep(300)
      assert {_pid, %{failures: 3}} = scheduler |> find_task(%{name: :failing})
    end

    test "tasks keep statistics of time to complete" do
      {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

      :timer.sleep(200)
      {_pid, %{last_completed_in: time}} = scheduler |> find_task(%{name: :tight_schedule})

      assert time > 0.0
    end
  end

  describe "failure handling" do
    test "when a process exceeeds :max_failures it is restarted" do
      capture_log(fn ->
        {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

        {pid, _} = scheduler |> find_task(%{name: :failing})
        assert Process.alive?(pid)

        :timer.sleep(500)

        # The old process is no longer alive
        refute Process.alive?(pid)
        {pid, _} = scheduler |> find_task(%{name: :failing})

        # A new task has been spawned
        assert Process.alive?(pid)
      end)
    end

    test "when a failing process is restarted, an error is logged" do
      log_output =
        capture_log(fn ->
          {:ok, _scheduler} = Subject.start_link(name: generate_name(Subject))
          :timer.sleep(500)
        end)

      assert log_output =~ ~r/scheduled task failed with:/
    end

    test "with unspecified :max_failures it is not restarted" do
      capture_log(fn ->
        {:ok, scheduler} = Subject.start_link(name: generate_name(Subject))

        {pid, _} = scheduler |> find_task(%{name: :failing_indefinitely})
        assert Process.alive?(pid)

        :timer.sleep(500)

        {^pid, _} = scheduler |> find_task(%{name: :failing_indefinitely})
        assert Process.alive?(pid)
      end)
    end
  end

  defp find_task(scheduler, %{name: name}) do
    scheduler
    |> Supervisor.which_children()
    |> Stream.map(fn {_, pid, _, _} -> {pid, :sys.get_state(pid)} end)
    |> Enum.find(fn {_, %{schedule: %{name: schedule_name}}} -> schedule_name == name end)
  end

  def generate_name(module) do
    :"#{module}-#{make_ref() |> inspect()}"
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :ex_schedule, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:ex_schedule, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/ex_scheduler.ex
defmodule ExScheduler do
  @moduledoc """
  Module providing a way to declare actions happening on an interval basis.

  Example

  ```
  defmodule Quiqup.Schedules.Developer do
    use ExScheduler

    schedule every: {6, :hours} do
      Developer.eat(:pizza)
    end

    schedule every: :hour, first_in: {20, :minutes} do
      Developer.drink(:coffee)
    end
  end
  ```

  Supported options:

  # Intervals
  `every`: The interval on which to run the task

  `first_in`: How much time to wait before the first run, defaults to 0

  Examples

  ```
  schedule every: :millisecond, do: Task.execute # every millisecond
  schedule every: :second, do: Task.execute      # every second
  schedule every: :minute, do: Task.execute      # every minute
  schedule every: :hour, do: Task.execute        # every hour
  schedule every: :day, do: Task.execute         # every day

  schedule every: {2, :milliseconds}, do: Task.execute # every 2 milliseconds
  schedule every: {2, :seconds}, do: Task.execute      # every 2 seconds
  schedule every: {2, :minutes}, do: Task.execute      # every 2 minutes
  schedule every: {2, :hours}, do: Task.execute        # every 2 hours
  schedule every: {2, :days}, do: Task.execute         # every 2 days

  schedule every: {2, :hours}, first_in: 0, do:
    Task.execute # every 2 hours first in 0 seconds
  end

  schedule every: {2, :hours}, first_in: {20, :minutes} do
    Task.execute # every 2 hours first in 20 minutes
  end
  ```

  # Failure handling
  `:max_failures` - Number of times to fail for the task process to be restarted, defaults to `:infinity`

  Examples

  ```
  schedule every: {2, :days}, max_failures: 5, do: something
  ```
  """

  defmacro __using__(_opts) do
    quote location: :keep do
      require Logger

      Module.register_attribute(__MODULE__, :schedules, accumulate: true)

      use Supervisor

      import ExScheduler

      @doc "Starts the Schedule with the given arguments"
      @spec start_link(list()) :: GenServer.on_start()
      def start_link(opts) do
        with namespace <- normalize_namespace(opts[:namespace]) do
          Supervisor.start_link(
            __MODULE__,
            put_in(opts[:namespace], namespace),
            name: name(opts[:name], namespace)
          )
        end
      end

      @doc false
      def init(opts) do
        schedules() |> Enum.map(&child_spec/1) |> supervise(strategy: :one_for_one)
      end

      @doc "Returns the namespace of the schedule"
      def namespace do
        self() |> Process.info([:links]) |> Access.get(:links) |> Enum.at(0) |> namespace
      end

      def namespace(server) when is_pid(server) do
        server
        |> Process.info([:registered_name])
        |> Access.get(:registered_name)
        |> namespace
      end

      def namespace(server) when is_atom(server) do
        size = __MODULE__ |> to_string |> byte_size

        case server |> to_string do
          <<_module::bytes-size(size)>> <> "." <> namespace -> normalize_namespace(namespace)
          _ -> nil
        end
      end

      ### Callbacks

      @doc false
      def handle_call(:state, _from, state), do: {:reply, state, state}

      defp normalize_namespace(ns) when is_bitstring(ns), do: String.to_atom(ns)
      defp normalize_namespace(ns), do: ns

      defp child_spec(schedule) do
        worker(ExScheduler.ScheduledTask, [schedule], id: schedule.id)
      end

      defp name(nil, nil), do: __MODULE__
      defp name(nil, ns), do: :"#{__MODULE__}.#{ns}"
      defp name(name, _namespace), do: name

      @before_compile ExScheduler
    end
  end

  @doc false
  defmacro __before_compile__(env) do
    quote do
      @doc "Returns the list of scheduled tasks"
      @spec schedules() :: list(map())
      def schedules, do: unquote(Module.get_attribute(env.module, :schedules))
    end
  end

  @doc "A macro that defines a recurrent task"
  @spec schedule(list(), list()) :: tuple()
  defmacro schedule(options, do: expression) do
    id = make_ref() |> inspect

    quote do
      def handle_task(unquote(id)), do: unquote(expression)

      @schedules Macro.escape(%{
                   id: unquote(id),
                   module: __MODULE__,
                   name: unquote(options)[:name],
                   max_failures: unquote(options)[:max_failures] || :infinity,
                   interval: ExScheduler.interval(unquote(options)[:every]),
                   first_in: ExScheduler.interval(unquote(options)[:first_in]) || 0
                 })
    end
  end

  def interval(n) when n in [nil, 0], do: 0
  def interval(n) when is_number(n), do: n * 1000
  def interval({_, :millisecond}), do: 1
  def interval({n, :milliseconds}), do: n
  def interval(:second), do: interval({1, :seconds})
  def interval(:minute), do: interval({1, :minutes})
  def interval(:hour), do: interval({1, :hours})
  def interval(:day), do: interval({24, :hours})

  def interval({1, duration}) when duration in [:second, :minute, :hour, :day] do
    apply(__MODULE__, :interval, [duration])
  end

  def interval({n, duration}) when duration in [:seconds, :minutes, :hours] do
    apply(:timer, duration, [n])
  end
end


lib/ex_schedule/scheduled_task.ex
defmodule ExSchedule.ScheduledTask do
  @moduledoc """
  Module responsible to isolate in a process, each recurring event declared
  with ExSchedule.Schedule.
  """

  defmodule State do
    @moduledoc false
    defstruct schedule: nil, runs: 0, failures: 0, total_failures: 0, last_completed_in: nil
  end

  use GenServer
  require Logger

  @doc "Starts a scheduled task process"
  @spec start_link(list()) :: GenServer.on_start()
  def start_link(schedule), do: GenServer.start_link(__MODULE__, schedule)

  @doc false
  def init(%{first_in: first_in} = schedule) do
    Process.send_after(self(), :run, first_in)

    {:ok, %State{schedule: schedule}}
  end

  @doc "Returns statistics for the given task process"
  @spec stats(pid()) :: %{runs: integer(), failures: integer(), last_completed_in: nil | number()}
  def stats(pid), do: GenServer.call(pid, :stats)

  # callbacks

  def handle_call(
        :stats,
        _from,
        %{
          schedule: %{name: name},
          runs: runs,
          failures: failures,
          last_completed_in: last_completed_in
        } = state
      ) do
    {:reply, %{name: name, runs: runs, failures: failures, last_completed_in: last_completed_in},
     state}
  end

  def handle_cast(:run, %{schedule: %{interval: interval} = schedule} = state) do
    with {:noreply, _} = ret <- schedule |> run |> handle_run(state) do
      Process.send_after(self(), :run, interval)
      ret
    end
  end

  def handle_info(:run, state), do: handle_cast(:run, state)

  defp handle_run({:ok, %{completed_in: time}}, %{runs: runs} = state) do
    {:noreply, %State{state | runs: runs + 1, failures: 0, last_completed_in: time}}
  end

  defp handle_run(
         {:error, error},
         %{schedule: %{module: mod, max_failures: max_failures}, failures: failures} = state
       )
       when failures + 1 >= max_failures do
    Logger.error(
      "[#{inspect(mod)}] scheduled task failed with: #{inspect(error)}, state: #{inspect(state)} "
    )

    {:stop, :normal, state}
  end

  defp handle_run(
         {:error, _error},
         %{runs: runs, failures: failures, total_failures: total_failures} = state
       ) do
    {:noreply,
     %State{state | runs: runs + 1, failures: failures + 1, total_failures: total_failures + 1}}
  end

  defp run(%{id: id, module: module} = _schedule) do
    fn -> apply(module, :handle_task, [id]) end |> timed_call
  end

  defp timed_call(f) do
    {:ok, %{completed_in: (f |> :timer.tc() |> elem(0)) / 1_000_000}}
  rescue
    e -> {:error, e}
  end
end


lib/ex_schedule.ex
defmodule ExSchedule do
  @moduledoc """
  Module providing a way to declare actions happening on an interval basis.

  Example

  ```
  defmodule Quiqup.Schedules.Developer do
    use ExSchedule

    schedule every: {6, :hours} do
      Developer.eat(:pizza)
    end

    schedule every: :hour, first_in: {20, :minutes} do
      Developer.drink(:coffee)
    end
  end
  ```

  Supported options:

  # Intervals
  `every`: The interval on which to run the task

  `first_in`: How much time to wait before the first run, defaults to 0

  Examples

  ```
  schedule every: :millisecond, do: Task.execute # every millisecond
  schedule every: :second, do: Task.execute      # every second
  schedule every: :minute, do: Task.execute      # every minute
  schedule every: :hour, do: Task.execute        # every hour
  schedule every: :day, do: Task.execute         # every day

  schedule every: {2, :milliseconds}, do: Task.execute # every 2 milliseconds
  schedule every: {2, :seconds}, do: Task.execute      # every 2 seconds
  schedule every: {2, :minutes}, do: Task.execute      # every 2 minutes
  schedule every: {2, :hours}, do: Task.execute        # every 2 hours
  schedule every: {2, :days}, do: Task.execute         # every 2 days

  schedule every: {2, :hours}, first_in: 0, do:
    Task.execute # every 2 hours first in 0 seconds
  end

  schedule every: {2, :hours}, first_in: {20, :minutes} do
    Task.execute # every 2 hours first in 20 minutes
  end
  ```

  # Failure handling
  `:max_failures` - Number of times to fail for the task process to be restarted, defaults to `:infinity`

  Examples

  ```
  schedule every: {2, :days}, max_failures: 5, do: something
  ```
  """

  defmacro __using__(_opts) do
    quote location: :keep do
      require Logger

      Module.register_attribute(__MODULE__, :schedules, accumulate: true)

      use Supervisor

      import ExSchedule

      @doc "Starts the Schedule with the given arguments"
      @spec start_link(list()) :: GenServer.on_start()
      def start_link(opts) do
        with namespace <- normalize_namespace(opts[:namespace]) do
          Supervisor.start_link(
            __MODULE__,
            put_in(opts[:namespace], namespace),
            name: name(opts[:name], namespace)
          )
        end
      end

      @doc false
      def init(opts) do
        schedules() |> Enum.map(&child_spec/1) |> Supervisor.init(strategy: :one_for_one)
      end

      @doc "Returns the namespace of the schedule"
      def namespace do
        self() |> Process.info([:links]) |> Access.get(:links) |> Enum.at(0) |> namespace
      end

      def namespace(server) when is_pid(server) do
        server
        |> Process.info([:registered_name])
        |> Access.get(:registered_name)
        |> namespace
      end

      def namespace(server) when is_atom(server) do
        size = __MODULE__ |> to_string |> byte_size

        case server |> to_string do
          <<_module::bytes-size(size)>> <> "." <> namespace -> normalize_namespace(namespace)
          _ -> nil
        end
      end

      ### Callbacks

      @doc false
      def handle_call(:state, _from, state), do: {:reply, state, state}

      defp normalize_namespace(ns) when is_bitstring(ns), do: String.to_atom(ns)
      defp normalize_namespace(ns), do: ns

      defp child_spec(schedule) do
        %{id: schedule.id, start: {ExSchedule.ScheduledTask, :start_link, [schedule]}}
      end

      defp name(nil, nil), do: __MODULE__
      defp name(nil, ns), do: :"#{__MODULE__}.#{ns}"
      defp name(name, _namespace), do: name

      @before_compile ExSchedule
    end
  end

  @doc false
  defmacro __before_compile__(env) do
    quote do
      @doc "Returns the list of scheduled tasks"
      @spec schedules() :: list(map())
      def schedules, do: unquote(Module.get_attribute(env.module, :schedules))
    end
  end

  @doc "A macro that defines a recurrent task"
  @spec schedule(list(), list()) :: tuple()
  defmacro schedule(options, do: expression) do
    id = make_ref() |> inspect

    quote do
      def handle_task(unquote(id)), do: unquote(expression)

      @schedules Macro.escape(%{
                   id: unquote(id),
                   module: __MODULE__,
                   name: unquote(options)[:name],
                   max_failures: unquote(options)[:max_failures] || :infinity,
                   interval: ExSchedule.interval(unquote(options)[:every]),
                   first_in: ExSchedule.interval(unquote(options)[:first_in]) || 0
                 })
    end
  end

  def interval(n) when n in [nil, 0], do: 0
  def interval(n) when is_number(n), do: n * 1000
  def interval({_, :millisecond}), do: 1
  def interval({n, :milliseconds}), do: n
  def interval(:second), do: interval({1, :seconds})
  def interval(:minute), do: interval({1, :minutes})
  def interval(:hour), do: interval({1, :hours})
  def interval(:day), do: interval({24, :hours})

  def interval({1, duration}) when duration in [:second, :minute, :hour, :day] do
    apply(__MODULE__, :interval, [duration])
  end

  def interval({n, duration}) when duration in [:seconds, :minutes, :hours] do
    apply(:timer, duration, [n])
  end
end
